modepy/modes.py

```diff
@@ -648,6 +662,41 @@ def tensor_product_basis(dims, basis_1d):
             _TensorProductBasisFunction(order, [basis_1d[i] for i in order])
             for order in gnitb(len(basis_1d), dims))
 
+
+def grad_tensor_product_basis(dims, basis_1d, grad_basis_1d):
+    """Provides derivatives for each of the basis functions generated by
+    :func:`tensor_product_basis`.
+
+    :returns: a :class:`tuple` of callables, where each one returns a
+        *dims*-dimensional :class:`tuple`, one for each derivative.
+
+    .. versionadded: 2020.3
+    """
+    from pytools import (
+            wandering_element,
+            generate_nonnegative_integer_tuples_below as gnitb)
+
+    func = (basis_1d, grad_basis_1d)
+    return tuple(
+            _TensorProductGradientBasisFunction(order, [
+                [func[i][k] for i, k in zip(iderivative, order)]
+                for iderivative in wandering_element(dims)
+                ])
+            for order in gnitb(len(basis_1d), dims))
+
+
+def legendre_tensor_product_basis(dims, order):
```

Comment:
Should these guys be exported in `modepy.__init__`? I mostly added them for testing to match the interface of `simplex_onb`.

Reply:
I think they're common enough that that makes sense.

Reply:
Done in 47a416f.